createkeydb.go

```diff
@@ -0,0 +1,292 @@
+package main
+
+import (
+	"os"
+	"errors"
+	"fmt"
+	"time"
+	"crypto/rsa"
+	"crypto/rand"
+	"crypto/sha1"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"path/filepath"
+	"math/big"
+
+	"github.com/urfave/cli"
+	"github.com/google/uuid"
+)
+
+func generate_rootCA(keyRepo, caType string, Organization []string) error {
+	var CN, location string
+	var expireDate time.Time
+	// Generate keypair
+	privkey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return err
+	}
+	err = privkey.Validate()
+	if err != nil {
+		return err
+	}
+
+	// Generate a self-signed certificate
+	serialNo, err := rand.Int(rand.Reader, new(big.Int).Lsh(big.NewInt(1), 128))
+	if caType == "sudi" {
+		CN = "Sudi rootCA"
+		location = filepath.Join(keyRepo, "sudiCA")
+		expireDate = time.Date(2099, time.December, 31, 23, 0, 0, 0, time.UTC)
+	} else if caType == "manifest" {
+		CN = "Manifest rootCA"
+		location = filepath.Join(keyRepo, "manifestCA")
+		expireDate = time.Now().AddDate(25,0,0)
+	} else {
+		return errors.New("unknown caType")
+	}
+
+	caTemplate := x509.Certificate {
+		SerialNumber:	serialNo,
+		Subject:		pkix.Name {
+							Organization: Organization,
+							OrganizationalUnit: []string{"PuzzleOS Machine Project snakeoil"},
+							CommonName: CN,
+						},
+		NotBefore:		time.Now(),
+		NotAfter:		expireDate,
+		IsCA:			true,
+		BasicConstraintsValid: true,
+	}
+
+	rootCA, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &privkey.PublicKey, privkey)
+	if err != nil {
+		return err
+	}
+	// Save the private key and cert to the specified directory
+	keyPEM, err := os.Create(filepath.Join(location, "key.pem"))
+    if err != nil {
+        return err
+    }
+    pkcs8, err := x509.MarshalPKCS8PrivateKey(privkey)
+    if err != nil {
```

Comment:
In this (and the next) case you are not closing keyPEM.

In theory golang will close it when it goes out of scope, but in practice (using a simple sleeping test) you can easily show that that doesn't happen very quickly.  The lack of 'defer keyPEM.Close()' at the top is in any case distracting so best to just have it there

Reply:
totally missed that! yes, let me fix it.